macros/shared/src/main/scala-2.x/zio/macros/AccessibleMacro.scala

```diff
@@ -153,40 +157,50 @@ private[macros] class AccessibleMacro(val c: Context) {
             if (isRepeatedParamType(arg)) q"${arg.name}: _*"
             else q"${arg.name}"
           })
-          q"_root_.zio.ZIO.accessM(_.get[Service].$name[..$typeArgs](...$argNames))"
+          q"_root_.zio.ZIO.accessM(_.get[Service[..$serviceTypeArgs]].$name[..$typeArgs](...$argNames))"
         case (_: Capability.Effect, _) =>
-          q"_root_.zio.ZIO.accessM(_.get[Service].$name)"
+          q"_root_.zio.ZIO.accessM(_.get[Service[..$serviceTypeArgs]].$name)"
         case (_: Capability.Managed, argLists) if argLists.flatten.nonEmpty =>
           val argNames = argLists.map(_.map { arg =>
             if (isRepeatedParamType(arg)) q"${arg.name}: _*"
             else q"${arg.name}"
           })
-          q"_root_.zio.ZManaged.service[Service].flatMap(_.$name[..$typeArgs](...$argNames))"
+          q"_root_.zio.ZManaged.service[Service[..$serviceTypeArgs]].flatMap(_.$name[..$typeArgs](...$argNames))"
         case (_: Capability.Managed, _) =>
-          q"_root_.zio.ZManaged.service[Service].flatMap(_.$name[..$typeArgs])"
+          q"_root_.zio.ZManaged.service[Service[..$serviceTypeArgs]].flatMap(_.$name[..$typeArgs])"
         case (_, argLists) if argLists.flatten.nonEmpty =>
           val argNames = argLists.map(_.map(_.name))
-          q"_root_.zio.ZIO.access(_.get[Service].$name[..$typeArgs](...$argNames))"
+          q"_root_.zio.ZIO.access(_.get[Service[..$serviceTypeArgs]].$name[..$typeArgs](...$argNames))"
         case (_, _) =>
-          q"_root_.zio.ZIO.access(_.get[Service].$name)"
+          q"_root_.zio.ZIO.access(_.get[Service[..$serviceTypeArgs]].$name)"
       }
 
-      if (isVal) q"$mods val $name: $returnType = $returnValue"
-      else
+      if (isVal && serviceTypeParams.isEmpty) q"$mods val $name: $returnType = $returnValue"
+      else {
+        val allTypeParams = serviceTypeParams ::: typeParams
+        val tagsForServiceTypeArgs = serviceTypeArgs.zipWithIndex.map {
+          case (name, idx) =>
+            val tagName = TermName(s"tag$idx")
+            q"$tagName: _root_.izumi.reflect.Tag[$name]"
```

Comment:
It is generally more performant to request the entire service tag. e.g.
instead of:
```scala
def v[T <: Foo: Tag, U >: Bar: Tag]: ZIO[Has[Module.Service[T, U]], Throwable, T]
```
->
```scala
def v[T <: Foo, U >: Bar](implicit ev: Tag[Module.Service[T, U]]): ZIO[Has[Module.Service[T, U]], Throwable, T]
```

In the latter case, if both T & U are statically known (their tags are not provided at runtime with Tag evidences), `Tag[Module.Service[T, U]]` will be a compile-time constant. In the former case there will always be a runtime cost associated with merging the runtime tags for `Module.Service`, `T` & `U` together.

Reply:
Thank you! Didn't think about that.